velox/connectors/hive/storage_adapters/s3fs/tests/S3InsertTest.cpp

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/init/Init.h>
+
+#include "gtest/gtest.h"
+#include "velox/common/file/FileSystems.h"
+#include "velox/connectors/hive/storage_adapters/s3fs/RegisterS3FileSystem.h"
+#include "velox/connectors/hive/storage_adapters/s3fs/tests/MinioServer.h"
+#include "velox/dwio/parquet/reader/ParquetReader.h"
+#include "velox/exec/TableWriter.h"
+#include "velox/exec/tests/utils/AssertQueryBuilder.h"
+#include "velox/exec/tests/utils/HiveConnectorTestBase.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::core;
+using namespace facebook::velox::exec;
+using namespace facebook::velox::exec::test;
+using namespace facebook::velox::connector;
+using namespace facebook::velox::connector::hive;
+using namespace facebook::velox::dwio::common;
+using namespace facebook::velox::test;
+using namespace facebook::velox::filesystems;
+
+class S3InsertTest : public testing::Test, public VectorTestBase {
+ public:
+  static constexpr char const* kMinioConnectionString{"127.0.0.1:7000"};
+  static void SetUpTestSuite() {
```

Comment:
I guess we use some of this boilerplate in some other test suites as well? If so, would it make sense to refactor this into a S3TestBase helper class?

Reply:
I simplified the `minioServer_` initialization a bit. We don't need the `nullptr` check as this is called globally only once for the entire test suite.
We use the gtest global `SetUpTestSuite()` in each test since we want to set up only once for the entire test suite.
Apart from this, each test has a slightly different `static void SetUpTestSuite()`. 
Since this is static, we cannot make it a virtual function in the base.

Reply:
Added a comment. To allow each test to have a separate Minio server, we need to ensure the ports are different and each test uses a separate temp directory. We can do both.

I chose the single-server approach because I wanted to minimize the test resource requirements. I achieved test isolation by ensuring each test used a separate bucket. This is error prone though.

I can open a separate PR with a Minio server for each test if you feel that is better.

Reply:
I'm asking because our default internal test runner, for example, runs different test cases in parallel in separate process, so this might not work due to port and file conflict. 

I recently added a small utility that returns an available port, could be helpful here.

Reply:
Makes sense. https://github.com/facebookincubator/velox/pull/4990 will be very helpful here.
I will open a separate PR once this PR lands and fix all the tests to use a base class.